pallets/accounts/src/lib.rs

```diff
@@ -993,5 +1169,201 @@ pub mod pallet {
+					);
+					log::trace!("Tickets to be withdrawn: {}", tickets_amount_to_withdraw);
+					log::trace!("NOVA to be withdrawn: {}", nova_amount_to_withdraw);
+
+					let tickets_amount_to_mint: <T as pallet_assets::Config>::Balance =
+						tickets_amount_to_withdraw.saturated_into();
+					let nova_amount_to_deposit: <T as pallet_balances::Config>::Balance =
+						nova_amount_to_withdraw.saturated_into();
+
+					ensure!(tickets_amount_to_withdraw > 0, Error::<T>::SystematicFailure);
+					ensure!(nova_amount_to_withdraw > 0, Error::<T>::SystematicFailure);
+
+					tickets_amount = tickets_amount.saturating_add(tickets_amount_to_mint);
+					nova_amount = nova_amount.saturating_add(nova_amount_to_deposit);
+
+					let nova_old_balance = pallet_balances::Pallet::<T>::free_balance(&account.clone());
+					ensure!(
+					nova_old_balance + nova_amount >=
+						<pallet_balances::Pallet<T> as Currency<T::AccountId>>::minimum_balance(
+						),
+					Error::<T>::SystematicFailure // this should never happen
+					);
+
+					if num_weeks_since_lock_frag == lock_period_in_weeks.clone() as u128 {
+						<EthLockedFrag<T>>::remove(eth_account.clone(), eth_lock.block_number);
+					} else {
+						// Update the week number of the latest withdraw
+						<EthLockedFrag<T>>::mutate(
+							&eth_account,
+							eth_lock.block_number,
+							|eth_lock| {
+								let eth_lock = eth_lock.as_mut().unwrap();
+								eth_lock.last_withdraw = num_weeks_since_lock_frag.clone();
+							},
+						);
+					}
+				}
+
+				// Checks completed, now write
+				// mint tickets
+				<pallet_assets::Pallet<T> as Mutate<T::AccountId>>::mint_into(
+					T::TicketsAssetId::get(),
+					&account.clone(),
+					tickets_amount,
+				)?;
+
+				// Assign NOVA
+				<pallet_balances::Pallet<T> as fungible::Mutate<T::AccountId>>::mint_into(
+					&account.clone(),
+					nova_amount,
+				)?;
+				Ok(())
+			} else {
+				return Err(Error::<T>::NothingToWithdraw.into());
+			}
+		}
+
+		fn initial_amount(amount: u128, percent: u128, current_frag_price: u128) -> u128 {
+			if amount == 0 {
+				return 0;
+			}
+			let percentage_amount = Self::apply_percent(amount, percent);
+			percentage_amount * current_frag_price * 100
+		}
+
+		/// Calculate a percentage
+		pub fn apply_percent(amount: u128, percent: u128) -> u128 {
+			if amount == 0 {
+				return 0;
+			}
+			amount * percent / 100
+		}
+
+		/// Get the price of FRAG from pallet-oracle
+		pub fn get_oracle_price() -> u128 {
+			1 // Assume the current price of 1 FRAG = 1 USD
+			 // TODO implement pallet Oracle
+		}
+
+		/// Convert the lock period integer retrieved from Ethereum event into the number of weeks.
+		/// Refer to https://github.com/fragcolor-xyz/hasten-contracts/blob/clamor/contracts/FragToken.sol
+		pub fn eth_lock_period_to_weeks(lock_period: u8) -> Result<u8, Error<T>> {
+			let sec = match lock_period {
+				0 => 2,  // 2 weeksu64
+				1 => 4,  // 1 month
+				2 => 12, // 3 months
```

Comment:
Probably doesn't matter too much but 3/6 months is 13/26 weeks respectively (would have to change the contract)
Feel free to leave as is

Reply:
good point 👍

Reply:
Fixing it thanks!